### Heroes Of Pymoli Data Analysis
* Of the 1163 active players, the vast majority are male (84%). There also exists, a smaller, but notable proportion of female players (14%).

* Our peak age demographic falls between 20-24 (44.8%) with secondary groups falling between 15-19 (18.60%) and 25-29 (13.4%).  
-----

### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data_df = pd.read_csv(file_to_load)

purchase_data_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [15]:
# test = grouped_purchase_data['Gender'].value_counts()
# test.index = ['Female','Male','Other / Non-Disclosed']
# test

In [112]:
# pd.options.display.max_rows = 10

## Player Count

* Display the total number of players


In [2]:
# Calculates and displays the total number of unique players

count_total_players = len(purchase_data_df['SN'].unique())

Count_total_players_df = pd.DataFrame({"Total Players": [count_total_players]})

# Display players count data frame
Count_total_players_df

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [3]:
# Identify unique set of items to get the count of items.
unique_items_count = len(purchase_data_df['Item ID'].unique())

# Average price of all purchases
average_price = purchase_data_df['Price'].mean(),

# Identify unique set of Purchase IDs to get the count of purchases.
purchase_count = len(purchase_data_df['Purchase ID'].unique())

# Total price of all the purchases
total_revenue = purchase_data_df['Price'].sum()

# Load all the above calculated result metrics into Purchasing Analysis (Total) summary data frame
purchase_analysis_total_summary_df = pd.DataFrame({"Number of Unique Items": unique_items_count,
                                                   "Average Purchase Price": average_price,
                                                   "Total Number of Purchases": purchase_count,
                                                   "Total Revenue": total_revenue
                                        })

# Format the Average Purchase Price and Total Revenue
purchase_analysis_total_summary_df['Average Purchase Price'] = purchase_analysis_total_summary_df['Average Purchase Price'].map('${:,.2f}'.format)
purchase_analysis_total_summary_df['Total Revenue'] = purchase_analysis_total_summary_df['Total Revenue'].map('${:,.2f}'.format)

# Display data frame
purchase_analysis_total_summary_df

Number of Unique Items Average Purchase Price  Total Number of Purchases  \
0                     179                  $3.05                        780   

  Total Revenue  
0     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [4]:
# Initial data is purchase data. To get the count of players by gender - create a new DF with only the required unique master data columns.
#    Then drop the duplicates from the newly created data frame.

unique_players_df = purchase_data_df[['SN','Gender', 'Age']].drop_duplicates().reset_index(drop=True)

# Calculate Gender Demographics

# Group by Gender to get counts of Male, Female and Other players
gender_summary_df = unique_players_df.groupby(by='Gender').count()

# Count and Percentage of players by Gender
gender_summary_df.rename(columns = {'SN': 'Total Count'}, inplace = 'True')
gender_summary_df['Percentage of Players'] = gender_summary_df['Total Count']/count_total_players*100
gender_summary_df = gender_summary_df[['Total Count', 'Percentage of Players']]

# Round and format the percentage value
gender_summary_df['Percentage of Players'] = gender_summary_df['Percentage of Players'].map('{:,.2f}%'.format)
gender_summary_df = gender_summary_df.sort_values(by=['Total Count'], ascending=False)

# Display summary Data Frame
gender_summary_df

Total Count Percentage of Players
Gender                                                  
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [5]:
# Purchasing Analysis by Gender.

# Group purchase data by Gender
grouped_purchase_data = purchase_data_df.groupby(['Gender'])

# Purchase count by Gender. This is not based on the unique players
purchase_count_by_gender = grouped_purchase_data['Purchase ID'].count()

# Average price by Gender. This is not based on the unique players
average_price = grouped_purchase_data['Price'].mean()

# Total Purchase value by Gender
total_price = grouped_purchase_data['Price'].sum()

# Average price by person. Average price is based on the unique players by Gender
average_price_per_person = total_price/gender_summary_df['Total Count']

# Load all the above calculated result metrics into Purchasing Analysis (Gender) summary data frame
purchase_analysis_gender_summary_df = pd.DataFrame({
                                            "Purchase Count": purchase_count_by_gender,
                                            "Average Purchase Price": average_price,
                                            "Total Purchase Value": total_price,
                                            "Avg Total Purchase per Person": average_price_per_person
                                          })

# Format the Average Purchase Price, Total Purchase Value and Average total Purchase per Person 
purchase_analysis_gender_summary_df['Average Purchase Price']=purchase_analysis_gender_summary_df['Average Purchase Price'].map('${:,.2f}'.format)
purchase_analysis_gender_summary_df['Total Purchase Value']=purchase_analysis_gender_summary_df['Total Purchase Value'].map('${:,.2f}'.format)
purchase_analysis_gender_summary_df['Avg Total Purchase per Person']=purchase_analysis_gender_summary_df['Avg Total Purchase per Person'].map('${:,.2f}'.format)

# Display Summary Data frame
purchase_analysis_gender_summary_df

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                             $1,967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [6]:
# Define bins for age groups
bins_to_cut = [0,9.9,14,19,24,29,34,39,100]

# Provide bin age group names
bin_group_names = ['<10', '10-14', '15-19', '20-24', '25-29', '30-34', '35-39', '40+']

In [7]:
# Age Demographics is based on the unique players
age_bins_df = pd.DataFrame(data=unique_players_df)

# Apply bins to the unique players data frame and categorize the players by age group
age_bins_df['Age Group'] = pd.cut(
                                    x=unique_players_df['Age'],
                                    bins=bins_to_cut,
                                    right=True,
                                    labels=bin_group_names
                                )

# Group by Age Groups
grouped_age_bins_df = age_bins_df.groupby(['Age Group'])

# Count the players by the Age Groups and calculate percentage of players in each Age group
bin_age_player_count = grouped_age_bins_df['SN'].count()
player_percentage = bin_age_player_count/count_total_players * 100

# Create the dataframe to store the player count and percentage by age group
age_demographics_df = pd.DataFrame({ 
                                    'Total Count': bin_age_player_count,
                                    'Percentage of Players': player_percentage
                                   })

# Round and Format the percentage value
age_demographics_df['Percentage of Players'] = age_demographics_df['Percentage of Players'].map('{:,.2f}%'.format)

# Display the data frame
age_demographics_df

Total Count Percentage of Players
Age Group                                   
<10                 17                 2.95%
10-14               22                 3.82%
15-19              107                18.58%
20-24              258                44.79%
25-29               77                13.37%
30-34               52                 9.03%
35-39               31                 5.38%
40+                 12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [8]:
# Purchasing Analysis is based on the purchase data
age_bins_df = pd.DataFrame(data=purchase_data_df)

# Apply bins to the purchase data and categorize the players by age group
age_bins_df['Age Group'] = pd.cut(
                                    x=purchase_data_df['Age'],
                                    bins=bins_to_cut,
                                    right=True,
                                    labels=bin_group_names
                                )
age_bins_df.head()

# Group by Age Groups
grouped_age_bins_df = age_bins_df.groupby(['Age Group'])

# Calculate Purchase count, Average Purchase price, Total Purchase Value and Average Total Purchase per Person
bin_age_count = grouped_age_bins_df['SN'].count()
average_price = grouped_age_bins_df['Price'].mean()
total_price = grouped_age_bins_df['Price'].sum()
average_price_per_person = total_price/age_demographics_df['Total Count']

# Populate the results into the Purchase analysis data frame
purchase_analysis_age_df = pd.DataFrame({ 'Purchase Count': bin_age_count,
                                         'Average Purchase Price': average_price,
                                         'Total Purchase Value': total_price,
                                         'Avg Total Purchase per Person': average_price_per_person
                                   })

# Round and Format the average and Total Purchase values.
purchase_analysis_age_df['Average Purchase Price'] = purchase_analysis_age_df['Average Purchase Price'].map('${:,.2f}'.format)
purchase_analysis_age_df['Total Purchase Value'] = purchase_analysis_age_df['Total Purchase Value'].map('${:,.2f}'.format)
purchase_analysis_age_df['Avg Total Purchase per Person'] = purchase_analysis_age_df['Avg Total Purchase per Person'].map('${:,.2f}'.format)

# Display the purchase analysis summary data frame
purchase_analysis_age_df

Purchase Count Average Purchase Price Total Purchase Value  \
Age Group                                                               
<10                    23                  $3.35               $77.13   
10-14                  28                  $2.96               $82.78   
15-19                 136                  $3.04              $412.89   
20-24                 365                  $3.05            $1,114.06   
25-29                 101                  $2.90              $293.00   
30-34                  73                  $2.93              $214.00   
35-39                  41                  $3.60              $147.67   
40+                    13                  $2.94               $38.24   

          Avg Total Purchase per Person  
Age Group                                
<10                               $4.54  
10-14                             $3.76  
15-19                             $3.86  
20-24                             $4.32  
25-29                             $3.81  
30-34                             $4.12  
35-39                             $4.76  
40+                               $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [9]:
# For identifying Top spenders, group Purchase data by player
top_spenders_calc_df = purchase_data_df.groupby(['SN'])

# Calculate Purchase counts, Average and Total price for each player
ts_purchase_count = top_spenders_calc_df['SN'].count()
ts_average_price = top_spenders_calc_df['Price'].mean()
ts_total_price = top_spenders_calc_df['Price'].sum()

# Populate the results into the top spenders dataframe
top_spenders_df = pd.DataFrame({'Purchase Count': ts_purchase_count,
                                'Average Purchase Price': ts_average_price,
                                'Total Purchase Value': ts_total_price
                               })

# Sort the dataframe by column 'Total Purchase Value' in descending order
top_spenders_df=top_spenders_df.sort_values(by=['Total Purchase Value'], ascending=False)

# Round and Format the Total and Average Purchase prices value
top_spenders_df['Average Purchase Price'] = top_spenders_df['Average Purchase Price'].map('${:,.2f}'.format)
top_spenders_df['Total Purchase Value'] = top_spenders_df['Total Purchase Value'].map('${:,.2f}'.format)

# Display preview of top spenders summary Data frame
top_spenders_df.head()

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.41               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [10]:
# Identify most popular items of purchase data by Item ID and Item name
most_popular_items_calc_df = purchase_data_df.groupby(['Item ID', 'Item Name'])

# Calculate the number of items sold, average item price and total item price
mpi_purchase_count = most_popular_items_calc_df['SN'].count()
mpi_item_price = most_popular_items_calc_df['Price'].mean()
mpi_total_price = most_popular_items_calc_df['Price'].sum()

# Populate the calculated results into the most popular items dataframe
most_popular_items_df = pd.DataFrame({'Purchase Count': mpi_purchase_count,
                                        'Item Price': mpi_item_price,
                                        'Total Purchase Value': mpi_total_price
                               })

# using a copy of this raw data frame to identify the most profitable item.
most_profitable_items_df = most_popular_items_df.copy()

# Sorting by column 'Total Purchase Count' of the item
most_popular_items_df=most_popular_items_df.sort_values(by=['Purchase Count'], ascending=False)

# Round and Format the Item and Total Purchase prices
most_popular_items_df['Item Price'] = most_popular_items_df['Item Price'].map('${:,.2f}'.format)
most_popular_items_df['Total Purchase Value'] = most_popular_items_df['Total Purchase Value'].map('${:,.2f}'.format)

# Display the summary of most popular items
most_popular_items_df.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [11]:
# Sory Item-wise purchase data by Total Purchase value to obtain the most profitable items
most_profitable_items_df = most_profitable_items_df.sort_values(by=['Total Purchase Value'], ascending=False)

# Round and Format the Item and Total Purchase prices
most_profitable_items_df['Item Price'] = most_profitable_items_df['Item Price'].map('${:,.2f}'.format)
most_profitable_items_df['Total Purchase Value'] = most_profitable_items_df['Total Purchase Value'].map('${:,.2f}'.format)

# Display the summary of most profitable items.
most_profitable_items_df.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

## Written Report

* Most number of players, approximately 45%, are in the age group 20-24; In the age groups less than 10 and above 40, there are only about 5% of players.

* Out of 576 players, 85% consist of male players.

* Top 3 spenders did not buy the most popular item "Final Critic". Top 3 most popular items were not bought by the top spender "Lisosia93".
